student/student_service.go

```diff
@@ -0,0 +1,143 @@
+package service
+
+import (
+    "context"
+    "errors"
+    "sync"
+    "time"
+
+    "github.com/google/uuid"
+    "example.com/student-service/proto" 
+    "google.golang.org/grpc/codes"
+    "google.golang.org/grpc/status"
+    "google.golang.org/protobuf/types/known/emptypb"
+    "google.golang.org/protobuf/types/known/timestamppb"
+)
+
+type Student struct {
+    ID        string
+    FirstName string
+    LastName  string
+    Grade     int32
+    CreatedAt time.Time
+}
+
+type Store struct {
+    mu       sync.RWMutex
+    students map[string]Student
+}
+
+type StudentServer struct {
+    proto.UnimplementedStudentServiceServer
+    store *Store
+}
+
+func NewStudentServer() *StudentServer {
+    return &StudentServer{
+        store: &Store{
+            students: make(map[string]Student),
+        },
+    }
+}
+
+func (s *StudentServer) CreateStudent(ctx context.Context, req *proto.CreateStudentRequest) (*proto.CreateStudentResponse, error) {
+    id := uuid.New().String()
+    now := time.Now()
+
+    s.store.mu.Lock()
+    s.store.students[id] = Student{
+        ID:        id,
+        FirstName: req.FirstName,
+        LastName:  req.LastName,
+        Grade:     req.Grade,
+        CreatedAt: now,
+    }
+    s.store.mu.Unlock()
+
+    return &proto.CreateStudentResponse{Id: id}, nil
+}
+
+func (s *StudentServer) GetStudent(ctx context.Context, req *proto.GetStudentRequest) (*proto.GetStudentResponse, error) {
+    s.store.mu.RLock()
+    student, ok := s.store.students[req.Id]
+    s.store.mu.RUnlock()
+
+    if !ok {
+        return nil, errors.New("student not found")
+    }
+
+    return &proto.GetStudentResponse{
+        Student: &proto.Student{
+            Id:        student.ID,
+            FirstName: student.FirstName,
+            LastName:  student.LastName,
+            Grade:     student.Grade,
+            CreatedAt: timestamppb.New(student.CreatedAt),
+        },
+    }, nil
+}
+
+func (s *StudentServer) UpdateStudent(ctx context.Context, req *proto.UpdateStudentRequest) (*emptypb.Empty, error) {
+    student := req.Student
+
+    s.store.mu.Lock()
+    defer s.store.mu.Unlock()
+    stored, ok := s.store.students[student.Id]
+    if !ok {
+        return nil, errors.New("student not found")
+    }
+
+    if student.CreatedAt != nil {
+        updatedCreatedAt := student.CreatedAt.AsTime()
+        if !updatedCreatedAt.Equal(stored.CreatedAt) {
+            return nil, errors.New("created_at field cannot be modified")
```

Comment:
grpc code should be appropriate at all errors that should be seen by the client!
errors.New could usually be the way to return internal error, for example in case you failed to setup connection with database you might return errors.New (Internal)